taf/testlib/common/loops.py

```diff
@@ -0,0 +1,303 @@
+                if do_sleep:
+                    time.sleep(cloop._interval)
+
+        cloop._before_next = f
+
+    add = add_interval
+
+    def __init__(self, interval):
+        super(CondLoopInterval, self).__init__()
+        self.add_interval(self, interval)
+
+    @property
+    def interval(self):
+        return self._interval
+
+
+class CondLoopCounter(CondLoopImpl):
+    """
+    @brief:  Add a loop counter end condition that fails after the amount of loops is reached.
+    """
+    OPTION = 'counter'
+
+    @classmethod
+    def add_counter(cls, cloop, count, *args, **kwargs):
+        cloop._count_max = count
+
+        def end_cond():
+            if not(cloop.loops_counter < cloop._count_max):
+                return True
+            return False
+
+        cloop.end_conds_fail.append(end_cond)
+
+    add = add_counter
+
+    def __init__(self, count):
+        super(CondLoopCounter, self).__init__()
+        self.add_counter(self, count)
+
+    @property
+    def count_max(self):
+        return self._count_max
+
+
+class CondLoopTimeout(CondLoopImpl):
+    """
+    @brief:  Add a timeout end condition that fails after the deadline has been reached.
+    """
+    OPTION = 'timeout'
+
+    @classmethod
+    def add_timeout(cls, cloop, timeout, *args, **kwargs):
+        cloop._start_time = int(time.time())
+        cloop._deadline = cloop._start_time + timeout
+
+        def end_cond():
+            if not(int(time.time()) < cloop._deadline):
+                return True
+            return False
+
+        cloop.end_conds_fail.append(end_cond)
+
+    add = add_timeout
+
+    def __init__(self, timeout):
+        super(CondLoopTimeout, self).__init__()
+        self.add_timeout(self, timeout)
+
+    @property
+    def start_time(self):
+        return self._start_time
+
+    @property
+    def deadline(self):
+        return self._deadline
+
+
+class ConditionalLoop(CondLoopImpl):
+    """
+    public API
+    """
+    BP_UNREG_OPTIONS = "Unregistered options detected({0}): {1}"
+    BP_UNREG_PAIR = "'{0}'='{1}'"
+
+    def __init__(self, **kwargs):
+        super(ConditionalLoop, self).__init__()
+
+        for cl_option, cl_type in self.OPTIONS.items():
+            with suppress(KeyError):
+                cl_value = kwargs.pop(cl_option)
+                cl_type.add(self, cl_value)
+
+        if kwargs:
+            raise CondLoopException(
+                self.BP_UNREG_OPTIONS.format(
+                    len(kwargs),
+                    ', '.join(self.BP_UNREG_PAIR.format(k, v) for k, v in kwargs.items())))
+
+
+# register default types automatically
+DEFAULT_TYPES = [CondLoopInterval, CondLoopCounter, CondLoopTimeout]
+for cl_type in DEFAULT_TYPES:
+    cl_type.register_option()
+
+
+# public helpers
+class GenericLoop(ConditionalLoop):
+    """
+    Generic loop helper
+    """
+    @property
+    def _fail_iter(self):
+        return (True for cond in self.end_conds_fail if cond())
```

Comment:
I would expect this to be:

`return (bool(cond()) for cond in self.end_cond_fail)`

so that each condition generates a result and then we can use `any` or `all` instead of `list` and thereby stop when the first "bad" condition is found, instead of trudging on through the rest.

And similar for `_succ_iter` below.